zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala

```diff
@@ -211,21 +214,28 @@ private[consumer] final class Runloop(
       ZIO.foreachDiscard(revokedStreams) { case (tp, control) =>
         val remaining = bufferedRecords.recs.getOrElse(tp, Chunk.empty)
 
-        control.finishWith(
-          remaining.map(
-            CommittableRecord(_, commit, getConsumerGroupMetadataIfAny)
+        ZIO.logDebug(s"Revoking topic-partition ${tp} with ${remaining.size} remaining records") *>
+          control.finishWith(
+            remaining.map(
+              CommittableRecord(_, commit, getConsumerGroupMetadataIfAny)
+            )
           )
-        )
       }
 
-    val acc = ChunkBuilder.make[Runloop.Request]()
+    val acc = ChunkBuilder.make[Request]()
     val buf = mutable.Map.empty[TopicPartition, Chunk[ByteArrayConsumerRecord]]
     buf ++= bufferedRecords.recs
     requests.foreach(req => if (isRevoked(req.tp)) buf -= req.tp else acc += req)
     val unfulfilledRequests = acc.result()
     val newBufferedRecords  = BufferedRecords.fromMutableMap(buf)
 
-    revokeAction.as(Runloop.RevokeResult(unfulfilledRequests, newBufferedRecords, assignedStreams))
+    val endRevokedRequests = ZIO.foreachDiscard(requests.filter(req => isRevoked(req.tp))) { req =>
+      ZIO.logTrace(s"Ending request for TP ${req.tp}") *> req.end.unit
+    }
+
+    endRevokedRequests *> revokeAction.as(
+      Runloop.RevokeResult(unfulfilledRequests, newBufferedRecords, assignedStreams)
+    )
```

Comment:
Not, sure if this is better:
```suggestion
    for {
      _ <- endRevokedRequests
      _ <- revokeAction
    } yield Runloop.RevokeResult(unfulfilledRequests, newBufferedRecords, assignedStreams)
```

Reply:
Or:
```suggestion
    endRevokedRequests *>
    revokeAction *>
    ZIO.succeed(
      Runloop.RevokeResult(unfulfilledRequests, newBufferedRecords, assignedStreams)
    )
```

(I don't like the `as` method as it hinders readability.)

Reply:
Feel free to ignore this...